internal/service/location/controller.go

```diff
@@ -109,30 +125,41 @@ func (c *Controller) UpdateLocation(ctx context.Context, request *locationV1beta
 			To:        to,
 		}
 
-		crossing, err := c.detector.DetectCrossing(ctx, movement)
+		detected, err := c.detector.DetectCrossing(ctx, movement)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("detect crossing error: %s", err)
 		}
-		if crossing != nil {
-			c.publishTollgateCrossing(ctx, crossing)
-			tollgateCrossing = &locationV1beta1.TollgateCrossing{
-				TollgateId: crossing.TollgateID,
-				Movement: &locationV1beta1.Movement{
-					FromLon: crossing.Movement.From.Lon,
-					FromLat: crossing.Movement.From.Lat,
-					ToLon:   crossing.Movement.To.Lon,
-					ToLat:   crossing.Movement.To.Lat,
-				},
-				Direction: string(crossing.Direction),
+		if detected != nil {
+			tollgateCrossing := crossing.NewTollgateCrossing(detected.TollgateID, movement.SubjectID, detected)
+			err := tollgateCrossing.Insert(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("crossing insert error: %s", err)
 			}
+			c.publishTollgateCrossing(ctx, detected)
+			return transformCrossing(tollgateCrossing), nil
 		}
 	}
+	return nil, nil
+}
 
-	return &locationV1beta1.UpdateLocationResponse{
-		WorkerId:         request.WorkerId,
-		TollgateCrossing: tollgateCrossing,
-		UpdateTime:       request.UpdateTime,
-	}, nil
+func transformCrossing(c *crossing.TollgateCrossing) *locationV1beta1.TollgateCrossing {
+	return &locationV1beta1.TollgateCrossing{
+		Id:         c.ID,
+		TollgateId: c.TollgateID,
+		WorkerId:   c.WorkerID,
+		Direction:  string(c.Crossing.Crossing.Direction),
+		Alg:        string(c.Crossing.Crossing.Alg),
+		Movement: &locationV1beta1.Movement{
+			FromLon: c.Crossing.Crossing.Movement.From.Lon,
+			FromLat: c.Crossing.Crossing.Movement.From.Lat,
+			ToLon:   c.Crossing.Crossing.Movement.To.Lon,
+			ToLat:   c.Crossing.Crossing.Movement.To.Lat,
```

Comment:
I'd prefer getters where possible. You get a lot of safety in terms of null panics.

Reply:
thought about that too,  when tollgate_crossing created/persisted Crossing struct is filled completely. so we are safe here, but overall such chaining smells and looks ugly. agreed.